jsonstar/JsonStar.Schema.Ast.Types.fst

```diff
@@ -0,0 +1,128 @@
+    | Date
+
+let format_type_to_string (x : format_type) = 
+    match x with
+    | Percentage -> "percentage"
+    | Date       -> "date"
+
+let format (f : format_type) : Tot unit = ()
+
+/// Attributes that can be associated with anything that represents a field in json-schema
+type value_attributes = 
+    {
+        _description  : option string;
+        _defaultValue : option string;
+        _readOnly     : option bool;
+        _format       : option format_type;
+    }
+
+/// Attributes that can be associated with anything that represents an object in json-schema
+type object_attributes =
+    {
+        _description : option string;
+        // Id, schemaVersion, title, AllowAdditionalProperties ... (?)
+    }
+
+type number_refinement = 
+    | Minimum : v:int -> number_refinement
+    | Maximum : v:int -> number_refinement
+
+type string_refinement = 
+    | MinLength : v:nat -> string_refinement
+    | MaxLength : v:nat -> string_refinement
+    | Pattern : v:string -> string_refinement
+
+type enum_refinement = 
+    | Allow : v:list string -> enum_refinement
+    | Disallow : v:list string -> enum_refinement
+
+type refinement_type =
+    | NumberRefinement : number_refinement -> refinement_type
+    | StringRefinement : string_refinement -> refinement_type
+    | EnumRefinement   : enum_refinement   -> refinement_type
+    // Some fields are restricted by (or calculated from) other fields
+    | Formula          : string            -> refinement_type
+    | And : refinement_type -> refinement_type -> refinement_type
+
+/// Primitive types
+type primitive = 
+    | Int :  primitive
+    //| Real : primitive
+    | String : primitive
+    | Boolean : primitive
+
+/// A single case discriminated union with a single field where the type name is equal to the case name
+/// NOTE: A common workaround to assign some metadata (via attributes) to a type abbreviation
+and typedef = 
+    {
+        _base : typ;
+        _attributes : value_attributes;
+    }
+
+and refinement = 
+    {
+        _base : typ;
+        _refinement : refinement_type;
+    }
+
+and enum = 
+    {
+        enum_values : list string;
+        enum_attributes : value_attributes;
+    }
+
+and array_type = 
+    | List : typ -> array_type
+    | Tuple : list typ -> array_type
+
+and array = 
+    {
+        // Refinements (like Min/Max items) included in the element_type
+        element_type : array_type;
+        _unique_items : option bool; // attributes?
+    }
+
+/// OneOf type is a discriminated union where each case has at most one field of type record
+// NOTE: An idea, a post processing step, detecting a refinement that restricts one_of to a particular case results in a flattening?
+and one_of = 
+    /// An extra enum field where its values are used to select a dependent subschema
+    | ByEnum : string -> enum -> list (string * option typ) -> one_of
+    // Add support for requiredWhen and similar
+
+and record_field = 
+    {
+        _name : string;
```

Comment:
Are all strings allowed as names? Is the "" (the empty string) a valid field name? Is `"5"` a valid field name? Stringly-typed code once again.

Reply:
This should be an easy one, as F* handles the name validation of types for me. On the other hand, I don't really need any extra guarantees on the name for now so perhaps not worth changing this yet.